nll type annotations in multisegment path

nikomatsakis opened the pull request:
This turned out to be sort of tricky. The problem is that if you have a path like

```
<Foo<&'static u32>>::bar
```

and it comes from an impl like `impl<T> Foo<T>` then the self-type the user gave doesn't *directly* map to the substitutions that the impl wants. To handle this, then, we have to preserve not just the "user-given substs" we used to do, but also a "user-given self-ty", which we have to apply later. This PR makes those changes.

It also removes the code from NLL relate-ops that handled canonical variables and moves to use normal inference variables instead. This simplifies a few things and gives us a bit more flexibility (for example, I predict we are going to have to start normalizing at some point, and it would be easy now).

r? @matthewjasper -- you were just touching this code, do you feel comfortable reviewing this?

Fixes #54574

nikomatsakis committed:
rename to `user_substs_applied_to_def` for consistency

nikomatsakis committed:
add `force_instantiate_unchecked` method

nikomatsakis committed:
make `TypeRelating` take an infcx again

At some point, I had thought to use this code to handle equality
comparisons for the `IfEq` verify bounds; at that point, we might not
have had an infcx to talk about. But we wound up doing "SCC
representatives" instead, so that's fine.

nikomatsakis committed:
rework `relate_type_and_user_type` to use type inference variables

We used to use a kind of "home-grown" variant where we tracked the
value of each canonical variable.

nikomatsakis committed:
move `force_instantiate_unchecked` to be local to `nll_relate` code

nikomatsakis committed:
introduce a `UserTypeAnnotation` enum

nikomatsakis committed:
introduce `FnDef` and `AdtDef` to `UserTypeAnnotation`

nikomatsakis committed:
introduce a more expressive `UserSubsts`

nikomatsakis committed:
pull the common code across user-ty variants up top

nikomatsakis committed:
handle user-self-type for def-ids

nikomatsakis committed:
avoid type variables in the self-type

nikomatsakis committed:
pass along `user_self_ty`

nikomatsakis committed:
update tests

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/441650379) ([raw log](https://api.travis-ci.org/v3/job/441650379/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

[00:04:12] travis_fold:start:tidy
travis_time:start:tidy
tidy check
[00:04:13] tidy error: /checkout/src/librustc_mir/hair/cx/expr.rs:725: line longer than 100 chars
[00:04:13] tidy error: /checkout/src/librustc/mir/visit.rs:217: line longer than 100 chars
[00:04:13] tidy error: /checkout/src/librustc/mir/visit.rs:786: line longer than 100 chars
[00:04:14] some tidy checks failed
[00:04:14] 
[00:04:14] 
[00:04:14] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:04:14] 
[00:04:14] 
[00:04:14] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:04:14] Build completed unsuccessfully in 0:00:46
[00:04:14] Build completed unsuccessfully in 0:00:46
[00:04:14] Makefile:79: recipe for target 'tidy' failed
[00:04:14] make: *** [tidy] Error 1

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:1c2510b8
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
---
travis_time:end:0c8fd2b9:start=1539609067031300881,finish=1539609067035337065,duration=4036184
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:200a90a8
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:02ceb97a
travis_time:start:02ceb97a
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:0beccaec
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

nikomatsakis committed:
pacify the mercilous tidy

(long lines)

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/441664051) ([raw log](https://api.travis-ci.org/v3/job/441664051/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:51:25] .................................................................................................... 2200/4605
[00:51:29] ..................i................................................................................. 2300/4605
[00:51:33] .................................................................................................... 2400/4605
[00:51:37] .................................................................................................... 2500/4605
[00:51:41] ...............................iiiiiiiii............................................................ 2600/4605
[00:51:47] .................................................................................................... 2800/4605
[00:51:52] .................................................................................................... 2900/4605
[00:51:55] ......................................................i............................................. 3000/4605
[00:51:58] .................................................................................................... 3100/4605
---
[01:05:25] thread 'main' panicked at 'Some tests failed', tools/compiletest/src/main.rs:501:22
[01:05:25] .F..............................................
[01:05:25] failures:
[01:05:25] 
[01:05:25] ---- [mir-opt] mir-opt/basic_assignment.rs stdout ----
[01:05:25] thread '[mir-opt] mir-opt/basic_assignment.rs' panicked at 'Did not find expected line, error: Mismatch in lines
[01:05:25] Current block: None
[01:05:25] Actual Line: "        AscribeUserType(_4, o, Ty(Canonical { variables: [], value: std::option::Option<std::boxed::Box<u32>> }));"
[01:05:25] Expected Line: "       AscribeUserType(_4, o, Canonical { variables: [], value: std::option::Option<std::boxed::Box<u32>> });"
[01:05:25] Test Name: rustc.main.SimplifyCfg-initial.after.mir
[01:05:25] ... (elided)
[01:05:25]    bb0: {
[01:05:25]    bb0: {
[01:05:25]        StorageLive(_1);
[01:05:25]        _1 = const false;
[01:05:25]        FakeRead(ForLet, _1);
[01:05:25]        StorageLive(_2);
[01:05:25]        StorageLive(_3);
[01:05:25]        _3 = _1;
[01:05:25]        _2 = move _3;
[01:05:25]        StorageDead(_3);
[01:05:25]        StorageLive(_4);
[01:05:25]        _4 = std::option::Option<std::boxed::Box<u32>>::None;
[01:05:25]        FakeRead(ForLet, _4);
[01:05:25]        AscribeUserType(_4, o, Canonical { variables: [], value: std::option::Option<std::boxed::Box<u32>> });
[01:05:25]        StorageLive(_5);
[01:05:25]        StorageLive(_6);
[01:05:25]        _6 = move _4;
[01:05:25]        replace(_5 <- move _6) -> [return: bb2, unwind: bb5];
[01:05:25] ... (elided)
[01:05:25]    bb2: {
[01:05:25]    bb2: {
[01:05:25]        drop(_6) -> [return: bb6, unwind: bb4];
[01:05:25] ... (elided)
[01:05:25]    bb5: {
[01:05:25]        drop(_6) -> bb4;
[01:05:25]    }
[01:05:25]    }
[01:05:25] Actual:
[01:05:25] fn main() -> (){
[01:05:25]     let mut _0: ();
[01:05:25]     scope 1 {
[01:05:25]         scope 3 {
[01:05:25]             scope 5 {
[01:05:25]                 scope 7 {
[01:05:25]                 scope 8 {
[01:05:25]                 scope 8 {
[01:05:25]                     let _5: std::option::Option<std::boxed::Box<u32>>;
[01:05:25]             }
[01:05:25]             scope 6 {
[01:05:25]             scope 6 {
[01:05:25]                 let _4: std::option::Option<std::boxed::Box<u32>> as (Ty(Canonical { variables: [], value: std::option::Option<std::boxed::Box<u32>> }), /checkout/src/test/mir-opt/basic_assignment.rs:37:18: 37:34);
[01:05:25]         }
[01:05:25]         scope 4 {
[01:05:25]         scope 4 {
[01:05:25]             let _2: bool;
[01:05:25]     }
[01:05:25]     scope 2 {
[01:05:25]     scope 2 {
[01:05:25]         let _1: bool;
[01:05:25]     }
[01:05:25]     let mut _3: bool;
[01:05:25]     let mut _6: std25]     }
[01:05:25]     bb8: {                              
[01:05:25]         StorageDead(_4);
[01:05:25]         StorageDead(_2);
[01:05:25]         StorageDead(_1);
[01:05:25]         return;
[01:05:25] }', tools/compiletest/src/runtest.rs:2932:13
[01:05:25] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:05:25] 
[01:05:25] 
[01:05:25] 
[01:05:25] failures:
[01:05:25]     [mir-opt] mir-opt/basic_assignment.rs
[01:05:25] 
[01:05:25] test result: FAILED. 47 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
[01:05:25] 
[01:05:25] 
[01:05:25] 
[01:05:25] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/mir-opt" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/mir-opt" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "mir-opt" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-5.0/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zunstable-options " "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "5.0.0\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
[01:05:25] 
[01:05:25] 
[01:05:25] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:05:25] Build completed unsuccessfully in 0:18:57
[01:05:25] Build completed unsuccessfully in 0:18:57
[01:05:25] make: *** [check] Error 1
[01:05:25] Makefile:58: recipe for target 'check' failed

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:08bd2490
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
---
travis_time:end:00cc59fb:start=1539614915685238599,finish=1539614915689433081,duration=4194482
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:1159acde
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:0985ca1e
travis_time:start:0985ca1e
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:05964000
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

nikomatsakis commented on src/librustc/ty/subst.rs:
```diff
@@ -564,3 +537,98 @@ impl<'a, 'gcx, 'tcx> SubstFolder<'a, 'gcx, 'tcx> {
+///
+/// when you then have a path like `<Foo<Vec<&'static u32>>>::method`,
+/// this struct would carry the def-id of the impl along with the
+/// self-type `Vec<u32>`. Then we can instantiate the parameters of
```
this is a bit messed up, yeah, I think the `Vec` is adding pretty little, actually.

nikomatsakis commented on src/librustc/infer/nll_relate/mod.rs:
```diff
@@ -0,0 +1,736 @@
+    fn push_outlives(&mut self, sup: ty::Region<'tcx>, sub: ty::Region<'tcx>);
+
+    /// Creates a new universe index. Used when instantiating placeholders.
+    fn next_subuniverse(&mut self) -> ty::UniverseIndex;
```
I did but that PR has not landed yet =)

nikomatsakis replied:
I guess it will land first though

nikomatsakis committed:
fix mir-opt test

bors commented:
:umbrella: The latest upstream changes (presumably #54858) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis committed:
extract `type_relate` into the inference context as `nll_relate`

The name is not great. Nor is the existence of this code great.  It
should be merged with the main "type relating code" at some point.

nikomatsakis committed:
make `instantiate_canonical_with_fresh_inference_vars` public again

This used to be public, then I made it private in a previous PR, but
there really wasn't a strong need for that.

matthewjasper reviewed (approved):
LGTM. Not sure how confident I am reviewing a PR of this size though.

nikomatsakis commented:
r? @pnkfelix

pnkfelix reviewed (commented):
If you end up rebasing this PR, you might as well move 24bee005b646692b265af026cf863acf9a322460 down next to e339e84fffb3156aaa78fbf686436de02ba4cf48 and then squash the former into the latter.

(But I haven't seen anything yet that would stop me from r+'ing this, I'm just noting it.)

pnkfelix reviewed (commented):
hooray for the new `UserTypeAnnotation` enum! I hope this will make the corresponding visit  methods easier to grok

pnkfelix reviewed (commented):
Eeek regarding e7ed997d87a1769bd68a86412ae2a417e4d36b99; its a shame we couldn't continue keeping the contents of `nll_relate` hidden away under `rustc_mir::borrow_check::nll` ...

pnkfelix commented on src/librustc/ty/subst.rs:
```diff
@@ -564,3 +537,98 @@ impl<'a, 'gcx, 'tcx> SubstFolder<'a, 'gcx, 'tcx> {
+    pub substs: &'tcx Substs<'tcx>,
+
+    /// The self-type, in the case of a `<T>::Item` path (when applied
+    /// to an inherent impl). See `UserSubsts` below.
```
I don't understand where the reader is meant to be led by the "See `UserSubsts` below."

pnkfelix replied:
(Probably a typo for `UserSelfTy`.)

nikomatsakis replied:
yes, fixing

pnkfelix commented on src/test/ui/nll/user-annotations/method-ufcs-inherent-1.rs:
```diff
@@ -0,0 +1,19 @@
+struct A<'a> { x: &'a u32 }
+
+impl<'a> A<'a> {
+    fn new<'b, T>(x: &'a u32, y: T) -> Self {
```
why is there a `'b` here?

pnkfelix replied:
(I see the same thing occuring in the other tests. To be honest I had thought that we didn't allow unreferenced lifetimes in generic binders ... but I guess that is just a lint, when it comes to lifetimes...?)

nikomatsakis replied:
copy and paste error, I think

pnkfelix commented on src/test/ui/nll/user-annotations/method-ufcs-inherent-1.rs:
```diff
@@ -0,0 +1,19 @@
+
+fn foo<'a>() {
+    let v = 22;
+    let x = A::<'a>::new(&v, 22);
```
I think you and I mirrored each other on our respective PR's; there should be an `//~ ERROR` on this line, right?

nikomatsakis replied:
Eep yes, will add.

pnkfelix commented:
I had a slew of nits. But none of them should block landing this PR, and frankly, I want to rebase my own work pronto on top of this.

pnkfelix commented:
@bors r+

bors commented:
:pushpin: Commit 7ce2e7a5faad9bb7fd2b49072c8d1399dc801933 has been approved by `pnkfelix`

<!-- @bors r=pnkfelix 7ce2e7a5faad9bb7fd2b49072c8d1399dc801933 -->

nikomatsakis commented:
@bors p=1 -- lots of stuff wants to build on this

nikomatsakis committed:
fix comment

nikomatsakis commented:
@pnkfelix 

> Eeek regarding e7ed997; its a shame we couldn't continue keeping the contents of nll_relate hidden away under rustc_mir::borrow_check::nll ...

Well, I was hoping they will get unified at some point, in which case it makes sense.

nikomatsakis committed:
add ~ERROR annotations

nikomatsakis commented:
@bors r=pnkfelix p=1

bors commented:
:pushpin: Commit b70b4a6814d420222f28684f9286a21ddb980bcf has been approved by `pnkfelix`

<!-- @bors r=pnkfelix b70b4a6814d420222f28684f9286a21ddb980bcf -->

bors commented:
:hourglass: Testing commit b70b4a6814d420222f28684f9286a21ddb980bcf with merge 01ca85becd45a4115bd5a1b367a1667c06f0906b...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/builds/19564467), [status-travis](https://travis-ci.org/rust-lang/rust/builds/442435710?utm_source=github_status&utm_medium=notification)
Approved by: pnkfelix
Pushing 01ca85becd45a4115bd5a1b367a1667c06f0906b to master...
